README.md
# SHRI homework. Infrastructure

Задачи создаются в очереди TMP
https://tracker.yandex.ru/TMP/order:updated:false/filter?resolution=empty()&author=codebor

## Bash
- chmod u+x release.sh
- you need to install jq `brew install jq`
- для запуска с секретами локально нужно создать файл `.env.local` по примеру `.env.local.example`
  и грепнуть их в скриптах
  ```bash
  OAuth=$(grep OAuth .env.local | cut -d '=' -f2)
  OrgId=$(grep OrgId .env.local | cut -d '=' -f2)
  ``` 
 
## Docker
You can build your container with 
```bash
docker build . -t shrihwinfrastructure
```
and run it with
```bash
docker run -p 3000:3000 shrihwinfrastructure
```

## Github Actions
### test.yml
  Запуск тестов без запуска дополнительных скриптов
### release.yml
  1. 📦 Install jq
    добавляем пакет jq для возможности вытаскивать ключи из json-ов в ответах
  2. 🎟️ Create ticket
    выполняем скрипт ./github/workflows/scripts/release.sh
  3. 📦 Install packages
    Устанавливаем пакеты с оптимизированными настройками
  4. 🧪 Run tests
    Запускаем тесты
### publish.yml
Сборка docker-образа и доставка до registry в Yandex.Cloud
1. Берем action https://github.com/yc-actions/yc-cr-login
2. Добавляем свои ключи в yc-sa-json-credentials
```bash
# инициализация аккаунта
# для полноценной настройки нужно создать пробный аккаунт
yc init
# необходимо создать сервисный аккаунт (см. документацию)
yc iam key create --service-account-name default-sa -o key.json
# получение реестра: crp00000000000000000
yc container registry create --name my-registry
```
см. https://nikolaymatrosov.medium.com/github-action-%D0%B4%D0%BB%D1%8F-%D0%BF%D1%83%D1%88%D0%B0-%D0%B2-yandex-cloud-container-registry-cbe91d8b0198


# Запуск на виртуальной машине
https://cloud.yandex.ru/docs/container-registry/solutions/run-docker-on-vm
1. Добавляем сервисному аккаунту роль container-registry.images.puller
  - я добавил все роли своему аккаунту codebor@yandex.ru
2. Создаем виртуальную машину
  - я создал ubuntu
3. Зашел на свою ВМ
```bash
echo <oauth-токен> | docker login --username oauth --password-stdin cr.yandex
ssh codebor@51.250.4.110
```
4. Поставил там докер `sudo apt install docker.io`
5. Выполнил команду
```bash
curl -H Metadata-Flavor:Google 169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token | cut -f1 -d',' | cut -f2 -d':' | tr -d '"' | sudo docker login --username iam --password-stdin cr.yandex
```
6. Закачал и установил свой контейнер из облака
```bash
sudo docker pull cr.yandex/crp8ambq9gr5dlh9u0sd/shrihwinfrastructure:17985340de0cb95f70dbfc9397cec6d08aa0323fc
sudo docker run -d -p 80:3000 cr.yandex/crp8ambq9gr5dlh9u0sd/shrihwinfrastructure:17985340de0cb95f70dbfc9397cec6d08aa0323fc
```
7. Теперь можно зайти в приложение так http://51.250.4.110/
  

CHANGELOG.md
# v1.1.1
6b8e3b4 - v1.1.1 (walborn, 3 minutes ago)
6908e49 - fix constatns (walborn, 4 minutes ago)
010a093 - I don't know why github actions don't see git tag list (walborn, 10 minutes ago)
222cb9f - asdfasdfasdg (walborn, 13 minutes ago)
695821c - asdfasdg (walborn, 16 minutes ago)
f807cd9 - v1.0.12 (walborn, 21 minutes ago)
4e8db3e - aldskfjasdlkfj (walborn, 22 minutes ago)
8827bc0 - void: changed log (walborn, 25 minutes ago)
232a2cf - void: change getting prev tag (walborn, 28 minutes ago)
105b1d1 - test (walborn, 35 minutes ago)
9ffcef9 - void: comment with tags by mask (walborn, 39 minutes ago)
4efffa3 - v1.0.7 (walborn, 74 minutes ago)
d82d90b - void: added tags comments (walborn, 74 minutes ago)
178d601 - void: added comment (walborn, 85 minutes ago)
6785e76 - Testing hw commit (Dmitriy, 7 hours ago)
763eb8e - v1.0.4 release (walborn, 7 hours ago)
a7e913b - fix: change queue in yandex tracker (walborn, 7 hours ago)
d03d649 - github actions: test release by tag (walborn, 12 hours ago)
dae95eb - test: fix index page (walborn, 12 hours ago)
65fca70 - Testing commit №2 (Dmitriy, 16 hours ago)
26b082f - Testing commit (Dmitriy, 16 hours ago)
500402e - v1.0.0 (Senior Pomidor, 21 hours ago)
b031c2e - documentation: deploy on yandex cloud (Senior Pomidor, 21 hours ago)
1985340 - fix dockerfile (Senior Pomidor, 23 hours ago)
d5cce50 - fix dockerfile (Senior Pomidor, 23 hours ago)
fd7a4d0 - fix dockerfile (Senior Pomidor, 23 hours ago)
33f45ea - fix yaml syntacsis and file paths (Senior Pomidor, 23 hours ago)
cc0a9b2 - added yandex cloud (Senior Pomidor, 23 hours ago)
fcd252b - added github actions (Senior Pomidor, 2 days ago)
7697b90 - added docker (Senior Pomidor, 6 days ago)
1fd5873 - added jest testing and bush scripts (Senior Pomidor, 7 days ago)
f5eabc3 - add request (Senior Pomidor, 7 days ago)
ddb02ae - init nextjs app (Senior Pomidor, 7 days ago)
# v1.0.7

# v1.0.7

# v1.0.7

# v1.0.7

# v1.0.7

# v1.0.7

# v1.0.7
4efffa3 - v1.0.7 (walborn, 23 minutes ago)
 d82d90b - void: added tags comments (walborn, 24 minutes ago)

# v1.0

# v1.0.7

# v1.0.4
763eb8e - v1.0.4 release (walborn, 6 hours ago)
 a7e913b - fix: change queue in yandex tracker (walborn, 6 hours ago)

# v1.0.4
763eb8e - v1.0.4 release (walborn, 6 hours ago)
 a7e913b - fix: change queue in yandex tracker (walborn, 6 hours ago)

# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
 dae95eb - test: fix index page (walborn, 4 hours ago)

# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago) dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago) dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago) dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 4 hours ago)
dae95eb - test: fix index page (walborn, 4 hours ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 68 minutes ago)
dae95eb - test: fix index page (walborn, 73 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 68 minutes ago)
dae95eb - test: fix index page (walborn, 73 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 67 minutes ago)
dae95eb - test: fix index page (walborn, 72 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 66 minutes ago)
dae95eb - test: fix index page (walborn, 71 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 65 minutes ago)
dae95eb - test: fix index page (walborn, 70 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 63 minutes ago)
dae95eb - test: fix index page (walborn, 68 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 62 minutes ago)
dae95eb - test: fix index page (walborn, 67 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 61 minutes ago)
dae95eb - test: fix index page (walborn, 66 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 60 minutes ago)
dae95eb - test: fix index page (walborn, 65 minutes ago)
# v1.0.3
d03d649 - github actions: test release by tag (walborn, 58 minutes ago)
dae95eb - test: fix index page (walborn, 63 minutes ago)
# v1.0.3

__test__/index.test.js
/**
 * @jest-environment jsdom
 */

 import React from 'react'
 import { render, screen } from '@testing-library/react'
 import IndexPage from '../pages/index'
 
 describe('Index page', () => {
   it('renders a button', () => {
     render(<IndexPage />)
 
     const button = screen.getByRole('button', {
       name: /myself/i,
     })
 
     expect(button).toBeInTheDocument()
   })
 })

.github/workflows/scripts/test.sh
#!/bin/bash
# OAuth and OrgId are taken from environment variables

issues="https://api.tracker.yandex.net/v2/issues/"

headerAuth="Authorization: OAuth $OAuth"
headerOrgId="X-Org-Id: $OrgId"
headerContentType="Content-Type: application/json"

tag=$(git describe --tags --abbrev=0)

taskId=$(curl --silent --location --request POST ${issues}_search \
  --header "$headerOrgId" \
  --header "$headerAuth" \
  --header "$headerContentType" \
  --data-raw '{
      "filter": {
        "unique": "'"codebor/${tag}"'"
      }
    }' | jq -r '.[0].key')


test=$(yarn test 2>&1 | tail -n +3 | tr -s "\n" " ")
result=${test##*Test Suites:}
commented=$(curl --location --request POST ${issues}${taskId}/comments \
--header "$headerAuth" \
--header "$headerOrgId" \
--header "$headerContentType" \
--data-raw '{
  "text": "'"$result"'"
}')

status=$(echo "$commented" | jq -r '.statusCode')

echo $commented
# if [ $status = 201 ]; then
#   echo "Added comment: TEST RESULT"
# elif [ $status = 404 ]; then
#   echo "Cannot add comment, task is not found"
# fi


testing.sh
#!/bin/bash
# OAuth и OrgId берутся из переменных окружения
OAuth=$(grep OAuth .env.local | cut -d '=' -f2)
OrgId=$(grep OrgId .env.local | cut -d '=' -f2)

issues="https://api.tracker.yandex.net/v2/issues/"


headerAuth="Authorization: OAuth $OAuth"
headerOrgId="X-Org-Id: $OrgId"
headerContentType="Content-Type: application/json"

tag=$(git describe --tags --abbrev=0)

taskId=$(curl --silent --location --request POST ${issues}_search \
  --header "$headerOrgId" \
  --header "$headerAuth" \
  --header "$headerContentType" \
  --data-raw '{
      "filter": {
        "unique": "'"${tag}"'"
      }
    }' | jq -r '.[0].key')

# todo - понять как проверять успешность тестов
validation=$(yarn test | tr -s "")


comment="Tests:\n $validation"

commented=$(curl --location --request POST ${issues}${taskId}/comments \
  --header "$headerOrgId" \
  --header "$headerAuth" \
  --header "$headerContentType" \
  --data-raw '{
    "text": "'"${comment}"'"
  }')


status=$(echo "$commented" | jq -r '.statusCode')
echo "Created new comment: $commented"

if [ $status = 201 ]; then
  echo "Added new comment TEST RESULT"
elif [ $status = 404 ]; then
  echo "Cannot add new comment, task is not found"
fi



.github/workflows/scripts/api/problem/update.sh
#!/usr/bin/env bash

# Запрос на поиск задачи
SEARCH_TASK=$(
  curl --location --request POST "$YT_HOST"'/v2/issues/_search' \
  --header 'Authorization: OAuth '"$YT_TOKEN" \
  --header 'X-Org-ID: '"$YT_ORG_ID" \
  --header 'Content-Type: application/json' \
  --data '{
    "filter": {
      "unique": "'"$YT_UNIQUE_PREFIX"'_'"$CURRENT_TAG_NAME"'"
    }
  }'
)

TASK_ID=$(echo "$SEARCH_TASK" | awk -F '"id":"' '{ print $2 }' | awk -F '","' '{ print $1 }')

# Запрос на обновление задачи
UPDATE_TASK_CODE_RESPONSE=$(
  curl -o /dev/null -s -w "%{http_code}\n" --location --request PATCH "$YT_HOST"'/v2/issues/'"$TASK_ID" \
  --header 'Authorization: OAuth '"$YT_TOKEN" \
  --header 'X-Org-ID: '"$YT_ORG_ID" \
  --header 'Content-Type: application/json' \
  --data "$REQUEST_DATA_TASK"
)

FIRST_CHAR_HTTP_RESPONSE=$(echo "$UPDATE_TASK_CODE_RESPONSE" | cut -c 1)

[ "$FIRST_CHAR_HTTP_RESPONSE" = "2" ]
exit $?

.github/workflows/scripts/release.sh
#!/bin/bash
# OAuth and OrgId are taken from environment variables

issues="https://api.tracker.yandex.net/v2/issues/"

headerAuth="Authorization: OAuth $OAuth"
headerOrgId="X-Org-Id: $OrgId"
headerContentType="Content-Type: application/json"

# detect current release tag number
tag=$(git describe --tags --abbrev=0)

author=$(git show $tag  --pretty=format:"Author: %an" --date=format:'%Y-%m-%d %H:%M:%S' --no-patch)
date=$(git show ${tag} | grep Date:)

# write changelog by commit history, from previous release tag till current
# prev=$(git describe --tags --abbrev=0 $tag^)
commits=$(git log --pretty=format:"%h - %s (%an, %ar)" | tr -s "\n")
echo -e "# $tag\n$commits\n$(cat CHANGELOG.md)" > CHANGELOG.md

created=$(curl --silent --location --request POST ${issues} \
--header "$headerAuth" \
--header "$headerOrgId" \
--header "$headerContentType" \
--data-raw '{
  "queue": "TMP",
  "summary": "'"codebor/$tag"'",
  "type": "task",
  "assignee": "codebor",
  "description": "'"$commits"'",
  "unique": "'"codebor/$tag"'"
}')

status=$(echo "$created" | jq -r '.statusCode')

if [ $status = 201 ]; then
  echo "Release created successfully"
elif [ $status = 404 ]; then
  echo "Not found"
elif [ $status = 409 ]; then
  echo "Cannot create task with the same release version"
  echo "Adding new comment then"

  taskId=$(curl --silent --location --request POST ${issues}_search \
    --header "$headerOrgId" \
    --header "$headerAuth" \
    --header "$headerContentType" \
    --data-raw '{
        "filter": {
          "unique": "'"codebor/${tag}"'"
        }
      }' | jq -r '.[0].key')

  updated=$(curl --location --request PATCH ${issues}${taskId} \
  --header "$headerAuth" \
  --header "$headerOrgId" \
  --header "$headerContentType" \
  --data-raw '{
    "summary": "'"codebor/$tag"'",
    "type": "task",
    "assignee": "codebor",
    "description": "'"$commits"'"
  }')

  status=$(echo "$updated" | jq -r '.statusCode')

  if [ $status = 200 ]; then
    echo "Task updated"
  elif [ $status = 404 ]; then
    echo "Task not found"
  elif [ $status = 422 ]; then
    echo "Unable to process the contained instructions"
  fi
fi
